### PR TITLE
[HAL][CUDA] Updates to cuda stream path to make it functional

### PR DESCRIPTION
**Pull request by ThomasRaoux**

### CODE REVIEW

File: iree/hal/cuda/stream_command_buffer.c [hunk 1]
```diff
@@ -202,14 +180,37 @@ static iree_status_t iree_hal_cuda_stream_command_buffer_fill_buffer(
...
+    case 1: {
+      CUDA_RETURN_IF_ERROR(
+          command_buffer->context->syms,
+          cuMemsetD8Async(dst, *(const uint8_t*)(pattern), num_elements,
+                          command_buffer->stream),
+          "cuMemsetD8Async");
+      break;
+    }
+    default:
+      assert(0 && "Unsupported element size");
```
Comment (benvanik) [inline iree/hal/cuda/stream_command_buffer.c:211]: return an error here instead

File: iree/hal/utils/deferred_command_buffer.c [hunk 2]
```diff
@@ -802,9 +803,28 @@ IREE_API_EXPORT iree_status_t iree_hal_deferred_command_buffer_apply(
   return status;
 }
 
+static iree_hal_command_buffer_mode_t iree_hal_deferred_command_buffer_mode(
```
Comment (benvanik) [inline iree/hal/utils/deferred_command_buffer.c:806]: move these up below _destroy (keeping the functions in the same order as the vtable makes it easier to scan the files when jumping around between implementations)

**Review by benvanik — APPROVED** [review]

File: iree/hal/cuda/stream_command_buffer.c (hunk 1 above)
Comment (ThomasRaoux) [reply iree/hal/cuda/stream_command_buffer.c]: done

File: iree/hal/utils/deferred_command_buffer.c (hunk 2 above)
Comment (ThomasRaoux) [reply iree/hal/utils/deferred_command_buffer.c]: done

**Pull request merged**